libfreerdp/core/nla.c

```diff
@@ -619,40 +814,231 @@ static int nla_client_init(rdpNla* nla)
+				 * hash identification in WinPR.
+				 */
+				nla->identity->PasswordLength += LB_PASSWORD_MAX_LENGTH;
+				usePassword = FALSE;
 			}
+		}
+
+		if (usePassword)
+		{
+			if (sspi_SetAuthIdentity(nla->identity, settings->Username, settings->Domain,
+			                         settings->Password) < 0)
+				return -1;
+		}
+	}
 
-			if (usePassword)
+	return TRUE;
+}
+
+#ifdef WITH_GSSAPI
+static const char* parseInt(const char* v, INT32* r)
+{
+	*r = 0;
+
+	/* check that we have at least a digit */
+	if (!*v || !((*v >= '0') && (*v <= '9')))
+		return NULL;
+
+	for ( ; *v && (*v >= '0') && (*v <= '9'); v++)
+	{
+		*r = (*r * 10) + (*v - '0');
+	}
+
+	return v;
+}
+
+static BOOL parseKerberosDeltat(const char* value, INT32* dest, const char* message)
+{
+	INT32 v;
+	const char* ptr;
+	/* determine the format :
+	 *   h:m[:s] (3:00:02) 			deltat in hours/minutes
+	 *   <n>d<n>h<n>m<n>s 1d4h	    deltat in day/hours/minutes/seconds
+	 *   <n>						deltat in seconds
+	 */
+	ptr = strchr(value, ':');
+	if (ptr)
+	{
+		/* format like h:m[:s] */
+		*dest = 0;
+		value = parseInt(value, &v);
+		if (!value || *value != ':')
+		{
+			WLog_ERR(TAG, "Invalid value for %s", message);
+			return FALSE;
+		}
+
+		*dest = v * 3600;
+
+		value = parseInt(value + 1, &v);
+		if (!value || (*value != 0 && *value != ':') || (v > 60))
+		{
+			WLog_ERR(TAG, "Invalid value for %s", message);
+			return FALSE;
+		}
+		*dest += v * 60;
+
+		if (*value == ':')
+		{
+			/* have optional seconds */
+			value = parseInt(value + 1, &v);
+			if (!value || (*value != 0) || (v > 60))
 			{
-				if (sspi_SetAuthIdentity(nla->identity, settings->Username, settings->Domain,
-				                         settings->Password) < 0)
-					return -1;
+				WLog_ERR(TAG, "Invalid value for %s", message);
+				return FALSE;
 			}
+			*dest += v;
+		}
+		return TRUE;
+	}
+
+	/* <n> or <n>d<n>h<n>m<n>s format */
+	value = parseInt(value, &v);
+	if (!value)
+	{
+		WLog_ERR(TAG, "Invalid value for %s", message);
+		return FALSE;
+	}
+
+	if (!*value || isspace(*value)) {
+		/* interpret that as a value in seconds */
+		*dest = v;
+		return TRUE;
+	}
+
+	*dest = 0;
+	do
+	{
+		INT32 factor;
+		INT32 maxValue;
+
+		switch(*value)
+		{
+		case 'd':
+			factor = 3600 * 24;
+			maxValue = 0;
+			break;
+		case 'h':
+			factor = 3600;
+			maxValue = 0;
+			break;
+		case 'm':
+			factor = 60;
+			maxValue = 60;
+			break;
+		case 's':
+			factor = 1;
+			maxValue = 60;
+			break;
+		}
+
+		if ((maxValue > 0) && (v > maxValue))
+		{
+			WLog_ERR(TAG, "invalid value for unit %c when parsing %s", *value, message);
+			return FALSE;
+		}
+
+		*dest += (v * factor);
+		value++;
+		if (!*value)
+			return TRUE;
+
+		value = parseInt(value, &v);
+		if (!value || !*value)
+		{
+			WLog_ERR(TAG, "Invalid value for %s", message);
+			return FALSE;
+		}
+
+	} while(TRUE);
+
+	return TRUE;
+}
+#endif
+
+static BOOL nla_setup_kerberos(rdpNla* nla)
+{
+#ifdef WITH_GSSAPI
```

Comment:
Can this guard be removed? don´t see anything `GSSAPI` specific here, or does it just fail in these cases?